core/src/main/java/cz/xtf/core/openshift/OpenShift.java

```diff
@@ -95,6 +103,13 @@ public class OpenShift extends DefaultOpenShiftClient {
     private static volatile String routeSuffix;
 
     public static final String KEEP_LABEL = "xtf.cz/keep";
+
+    /**
+     * Used to cache created Openshift clients for given test case.
+     */
+    public static final Multimap<String, OpenShift> testcaseToOpenshiftClientMap = Multimaps
```

Comment:
Test Case Name is part of OpenShiftConfig, right? Don't we want Map<OpenShiftConfig, OpenShift> cache?

Reply:
I don't think it would be good to use `OpenShiftConfig` as key here as most instances of `OpenShiftConfig` would differ just by namespace which is retrieved from testcase->namespace map.

Reply:
> most instances of OpenShiftConfig would differ just by namespace
Exactly. 

And I thought I can't reuse OpenShift Client between different testcases(namespaces), because OpenShiftClient usage relies on default namespace which it was created with . We cannot switch namespace in runtime.

Reply:
I've changed the logic here. I see using testcase-> openshift is not effective when there is single namespace which is default use as it would cache as many clients as there are test cases. 
So ideal would be to have simple Set<OpenShift> but this on the other hand would not be effective for this feature where each testcase has different namespace so each OpenShift instance creation would iterate over this set.

Having Map<OpenShiftConfig, OpenShift> cache will not work as namespace in `OpenShiftConfig` is calculated dynamically (based on currently running testcase) but having Map<namespace, OpenShift> would work. 

It would be effective for both of the cases above. I'll re-push the logic.

Reply:
> Having Map<OpenShiftConfig, OpenShift> cache will not work as namespace in OpenShiftConfig is calculated dynamically (based on currently running testcase) but having Map<namespace, OpenShift> would work.

I don't get that argument. Isn't `namespace` also calculated dynamically based on currently running testcase?

Does `Map<namespace, OpenShift>` mean I can have just one OpenShift client for one namespace?

Reply:
`I don't get that argument. Isn't namespace also calculated dynamically based on currently running testcase?`

The reason why `OpenShiftConfig` cannot be used as key in map is that `OpenShiftConfig` instance is calculating namespace dynamically which is derived from currently running testcase name. So once a few test cases are executed it would map to `OpenShift` instances with different namespace than calculated in `OpenShiftConfig`. 

`Does Map<namespace, OpenShift> mean I can have just one OpenShift client for one namespace?`

It's `MultiMap` so one key can have multiple values. This mapping will group all openshift instances (usually the difference will be just whether it's for "master() or admin()") with the same namespace keeping number of newly created and cached `OpenShift` instances at minimum.

Reply:
>  So once a few test cases are executed it would map to OpenShift instances with different namespace than calculated in OpenShiftConfig.

Please could you leverage more which condition you have in mind? It is still not clear to me. 

I am thinking of 1. paralel run 2. xtf.openshift.namespace.per.testcase=true 3. xtf.openshift.namespace.per.testcase=false, but could not see the problem.

Reply:
> Please could you leverage more which condition you have in mind? It is still not clear to me.

I think `Map<OpenShiftConfig, OpenShift>` would not work. Let's imagine that that Test1 is running in namespace-test1 and openShiftConfig1 -> openshift1 is put into the this map. openShiftConfig1 does not contain variable to store namespace. There is just static `namespace()` method which returns namespace based on currently running test case.  Then Test2 is started and there is openShiftConfig2. Problem I see here is that openShiftConfig2 is equal to openShiftConfig1. So when openshift2 should be created with namespace-test2 then it will actually return cached openshift1 client.

Reply:
> openShiftConfig1 does not contain variable to store namespace

Ok, now I see what you mean. 

Alternativelly we can add fields to `OpenShiftConfig`(But it seems it is not how whole package is designed). Or alternativelly use `io.fabric8.openshift.client.OpenShiftConfig` as a key (Not sure if consfusion does not comes there exists same classes in different packages)?

[1] https://github.com/mnovak1/xtf/blob/ff552bba7aef990e3e1f7885d5c1f343d87f44e4/core/src/main/java/cz/xtf/core/openshift/OpenShift.java#L1384

Reply:
Can it be set as resolved?